src/app.rs

```diff
@@ -245,6 +270,32 @@ impl App {
         Ok(InclusionProofResponse(proof))
     }
 
+    /// # Errors
+    ///
+    /// Will return `Err` if the provided proof is invalid.
+    #[instrument(level = "debug", skip_all)]
+    pub async fn verify_semaphore_proof(
+        &self,
+        request: &VerifySemaphoreProofRequest,
+    ) -> Result<VerifySemaphoreProofResponse, ServerError> {
+        let Some(root_state) = self.database.get_root_state(&request.root).await? else {
+            return Ok(VerifySemaphoreProofResponse::InvalidRoot)
+        };
+
+        let checked = verify_proof(
+            request.root,
+            request.nullifier_hash,
+            request.signal_hash,
+            request.external_nullifier_hash,
+            &request.proof,
+        );
+        match checked {
+            Ok(true) => Ok(VerifySemaphoreProofResponse::Ok(root_state)),
+            Ok(false) => Ok(VerifySemaphoreProofResponse::InvalidProof),
+            Err(_) => Err(ServerError::ProverError),
```

Comment:
Can we log the error here for visibility? It could happend pretty often so maybe on `debug` level?